packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js

```diff
@@ -0,0 +1,363 @@
+          .connect(roles.randomPerson)
+          .setBeaconIdWithExpirationTimestamp(randomBeaconId, now + 10)
+      ).to.be.revertedWith('Ownable: caller is not the owner');
+    });
+  });
+});
+
+describe('setBeaconIdWithIndefiniteWhitelistStatus', function () {
+  context('caller is owner', function () {
+    it('sets the indefinite whitelist status for a beaconId', async function () {
+      const randomBeaconId = hre.ethers.utils.randomBytes(32);
+      expect(await selfServeRrpBeaconServerWhitelister.beaconIdToIndefiniteWhitelistStatus(randomBeaconId)).to.equal(
+        false
+      );
+      await expect(
+        selfServeRrpBeaconServerWhitelister
+          .connect(roles.manager)
+          .setBeaconIdWithIndefiniteWhitelistStatus(randomBeaconId, true)
+      ).to.emit(selfServeRrpBeaconServerWhitelister, 'SetBeaconIdWithIndefiniteWhitelistStatus');
+
+      expect(await selfServeRrpBeaconServerWhitelister.beaconIdToIndefiniteWhitelistStatus(randomBeaconId)).to.equal(
+        true
+      );
+    });
+  });
+  context('caller is not owner', function () {
+    it('reverts', async function () {
+      const randomBeaconId = hre.ethers.utils.randomBytes(32);
+      await expect(
+        selfServeRrpBeaconServerWhitelister
+          .connect(roles.randomPerson)
+          .setBeaconIdWithIndefiniteWhitelistStatus(randomBeaconId, true)
+      ).to.be.revertedWith('Ownable: caller is not the owner');
+    });
+  });
+});
+
+describe('whitelistReaderWithExpiration', function () {
+  context('beaconId expirationTimestamp is valid', function () {
+    it('whitelists a reader', async function () {
+      await expect(rrpBeaconServer.connect(roles.beaconReader).readBeacon(beaconId)).to.be.revertedWith(
+        'Sender not whitelisted'
+      );
+
+      let whitelistStatus = await rrpBeaconServer.beaconIdToReaderToWhitelistStatus(
+        beaconId,
+        roles.beaconReader.address
+      );
+      expect(whitelistStatus.expirationTimestamp).to.equal(0);
+      expect(whitelistStatus.indefiniteWhitelistCount).to.equal(0);
+      await expect(
+        selfServeRrpBeaconServerWhitelister.whitelistReaderWithExpiration(beaconId, roles.beaconReader.address)
+      )
+        .to.emit(selfServeRrpBeaconServerWhitelister, 'WhitelistedReaderWithExpiration')
+        .withArgs(beaconId, roles.beaconReader.address);
+      whitelistStatus = await rrpBeaconServer.beaconIdToReaderToWhitelistStatus(beaconId, roles.beaconReader.address);
+      expect(whitelistStatus.expirationTimestamp).to.equal(beaconIdExpirationTimeStamp);
+      expect(whitelistStatus.indefiniteWhitelistCount).to.equal(0);
+
+      const beaconResponse = await rrpBeaconServer.connect(roles.beaconReader).readBeacon(beaconId);
+      expect(beaconResponse.value).to.equal(123);
+      expect(beaconResponse.timestamp).to.equal(encodedTimestamp);
+    });
+  });
+  context('beaconId expirationTimestamp is invalid', function () {
+    it('reverts', async function () {
```

Comment:
Shouldn't we have a similar test that tries to whitelist a reader, but doesn't have the permissions for it?

Reply:
do you mean the contract itself doesn't have the permissions ?